### DISCUSSION TITLE
How to achieve accurate rate-limits on Kong CE db-less mode?

### DISCUSSION THREAD

**Discussion by amitonlentra** [Kong/kong#13510] [Help]
We deploy kong 3.1.1 on kubernetes through helm. We are currently running the community edition in db-less mode. There are multiple instances of kong configured to achieve high availability and scalability. Kong enables us with routing, rate-limiting, authentication (custom plugin) and monitoring. We have rate-limits enabled on different routes across multiple services. Our main concern is having accurate rate limits.

We are evaluating multiple approaches to achieve this

1. Migrating to postgres db - This could turn out to be a performance bottleneck given that every api call would hit the database
2. Integrating redis - Given that the CE does not allow connecting to a redis cluster, a standalone redis node would be a single point of failure
3. Load balancer - There is mention about consistent hashing [here](https://docs.konghq.com/hub/kong-inc/rate-limiting/#backend-protection) that should be configured to direct the user to the same kong node. How do we configure this in a kubernetes service?

Are there any other approaches that we could explore?

Thanks

**Comment by tobiasehlert** [comment]
Reasonable question @amitonlentra, but I guess that it's like that in the CE-version.
I use the cache-proxy plugin and then caching is limited to that particular pods memory, since Redis is a EE feature there.
I guess it's the same approach here, so either upgrade to EE to support a Redis cluster or write a custom plugin.

The approach 3 with load balancer, take a look at the "sticky sessions" feature available in some. That way you could let a user return to the same Kong node, but keep in mind that this could result in uneven traffic.

**Comment by amitonlentra** [comment]
> I use the cache-proxy plugin and then caching is limited to that particular pods memory

Why do we need to explicitly use a caching proxy plugin since it anyways
uses pods memory? We are anyways running with db-less mode.

> The approach 3 with load balancer, take a look at the "sticky sessions" feature available in some.

We anyways have nginx in front of kong which must be having this feature.

Thanks for the valuable inputs.

**Comment by jeremyjpj0916** [comment]
Would recommend yall look here for dbless but also redis based on CE: https://github.com/dream11/kong-scalable-rate-limiter

**Reply by tobiasehlert** [reply]
Thanks @jeremyjpj0916 👍
